server/src/internalClusterTest/java/org/opensearch/search/profile/aggregation/AggregationProfilerIT.java

```diff
@@ -1061,15 +1061,27 @@ public void testTopHitsAggregationFetchProfiling() throws Exception {
                 assertTrue("Should be top_hits aggregation fetch", topHitsFetch1.getQueryName().startsWith("fetch_top_hits_aggregation"));
                 assertTrue("Should contain aggregation name", topHitsFetch1.getQueryName().contains("top_hits_agg1"));
                 assertNotNull(topHitsFetch1.getTimeBreakdown());
-                assertEquals("Top hits fetch should have 1 child (FetchSourcePhase)", 1, topHitsFetch1.getProfiledChildren().size());
-                assertEquals("FetchSourcePhase", topHitsFetch1.getProfiledChildren().get(0).getQueryName());
+                // With concurrent segment search, profiled children might be empty or have different structure
+                if (!topHitsFetch1.getProfiledChildren().isEmpty()) {
+                    // Verify at least one child is FetchSourcePhase when children exist
```

Comment:
I am wondering in what cases `topHitsFetch1.getProfiledChildren()` could be empty? While the structure could be different, but why could it be empty?

Reply:
The children can be empty due to a race condition in profile collection with concurrent segment search.
When concurrent search is enabled, FlatFetchProfileTree uses thread IDs to track phases (element + "_" + Thread.currentThread().getId()). This can cause sub-phases to not link correctly to their parent in the profile tree due to timing issues.

The fetch phase still executes FetchSourcePhase correctly, but the profiling infrastructure doesn't always capture the parent-child relationship.

Reply:
> The fetch phase still executes FetchSourcePhase correctly, but the profiling infrastructure doesn't always capture the parent-child relationship.

This seems more like a bug to me, than a flaky test. I am wondering if this issue is fixable and we should instead have a github issue to address it?

`ConcurrentQueryProfileBreakdown` has `associateCollectorToLeaves` method to build the relationship correctly instead of relying on the threadId. Maybe we can use the same logic here as well?

Reply:
okay, after spending another hour or so, I just realized that the maps need to be synchronized to avoid running into this issue https://github.com/opensearch-project/OpenSearch/pull/19164. After enabling concurrent implementations, I could get through more than 550 executions without running into failure. Previously, I could barely hit 200 executions